paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/commands/WriteIntoPaimonTable.scala

```diff
@@ -0,0 +1,199 @@
+    val bucketColIdx = withBucketCol.schema.size - 1
+    val withBucketDataEncoder = RowEncoder.apply(withBucketCol.schema).resolveAndBind()
+    val toRow = withBucketDataEncoder.createSerializer()
+    val fromRow = withBucketDataEncoder.createDeserializer()
+
+    val bucketProcessor = if (isDynamicBucketTable) {
+      DynamicBucketProcessor(table, rowType, bucketColIdx, numSparkPartitions, toRow, fromRow)
+    } else {
+      CommonBucketProcessor(writeBuilder, bucketColIdx, toRow, fromRow)
+    }
+
+    val committables =
+      withBucketCol
+        .mapPartitions(bucketProcessor.processPartition)(withBucketDataEncoder)
+        .toDF()
+        .repartition(partitionCols ++ Seq(col(BUCKET_COL)): _*)
+        .mapPartitions {
+          iter =>
+            val write = writeBuilder.newWrite()
+            try {
+              iter.foreach {
+                row =>
+                  val bucket = row.getInt(bucketColIdx)
+                  val bucketColDropped = originFromRow(toRow(row))
+                  write.write(new DynamicBucketRow(new SparkRow(rowType, bucketColDropped), bucket))
+              }
+              val serializer = new CommitMessageSerializer
+              write.prepareCommit().asScala.map(serializer.serialize).toIterator
+            } finally {
+              write.close()
+            }
+        }
+        .collect()
+        .map(deserializeCommitMessage(serializer, _))
+
+    try {
+      val tableCommit = writeBuilder.newCommit().asInstanceOf[InnerTableCommit]
+      tableCommit.commit(committables.toList.asJava)
+    } catch {
+      case e: Throwable => throw new RuntimeException(e);
+    }
+
+    Seq.empty
+  }
+
+  override def withNewChildrenInternal(newChildren: IndexedSeq[LogicalPlan]): LogicalPlan =
+    this.asInstanceOf[WriteIntoPaimonTable]
+}
+
+object WriteIntoPaimonTable {
+
+  sealed trait BucketProcessor {
+    def processPartition(rowIterator: Iterator[Row]): Iterator[Row]
+  }
+
+  case class CommonBucketProcessor(
+      writeBuilder: BatchWriteBuilder,
+      bucketColIndex: Int,
+      toRow: ExpressionEncoder.Serializer[Row],
+      fromRow: ExpressionEncoder.Deserializer[Row])
+    extends BucketProcessor {
+
+    private val rowType = writeBuilder.rowType
+
+    def processPartition(rowIterator: Iterator[Row]): Iterator[Row] = {
+      val batchTableWrite = writeBuilder.newWrite()
+      new Iterator[Row] {
+        override def hasNext: Boolean = rowIterator.hasNext
+
+        override def next(): Row = {
+          val row = rowIterator.next
+          val sparkInternalRow = toRow(row)
+          sparkInternalRow.setInt(
+            bucketColIndex,
+            batchTableWrite.getBucket(new SparkRow(rowType, row)))
+          fromRow(sparkInternalRow)
+        }
+      }
+    }
+  }
+
+  case class DynamicBucketProcessor(
+      fileStoreTable: FileStoreTable,
+      rowType: RowType,
+      bucketColIndex: Int,
+      numSparkPartitions: Long,
```

Comment:
Maybe we should use `TaskContext.numPartitions`?

Reply:
`TaskContext.get.numPartitions` is a new api in Spark 3.4.0. If use this and , maybe we have to make compatible with other spark versions later.